ddl/placement/bundle.go

```diff
@@ -49,6 +49,224 @@ func NewBundle(id int64) *Bundle {
 	}
 }
 
+// NewBundleFromConstraintsOptions will transform constraints options into the bundle.
+func NewBundleFromConstraintsOptions(constraints string, leaderConstraints string, learnerConstraints string, followerConstraints string, voterConstraints string, followerCount uint64, voterCount uint64, learnerCount uint64) (*Bundle, error) {
+	if voterCount+followerCount == 0 {
+		return nil, fmt.Errorf("%w: not enough voter, please specify voter or follower count", ErrInvalidPlacemeOptions)
+	}
+
+	CommonConstraints, err := NewConstraintsFromYaml([]byte(constraints))
+	if err != nil {
+		return nil, fmt.Errorf("%w: 'Constraints' should be [constraint1, ...] or any yaml compatible array representation", err)
+	}
+
+	LeaderConstraints, err := NewConstraintsFromYaml([]byte(leaderConstraints))
+	if err != nil {
+		return nil, fmt.Errorf("%w: 'LeaderConstraints' should be [constraint1, ...] or any yaml compatible array representation", err)
+	}
+	for _, cnst := range CommonConstraints {
+		if err := LeaderConstraints.Add(cnst); err != nil {
+			return nil, fmt.Errorf("%w: LeaderConstraints conflicts with Constraints", err)
+		}
+	}
+
+	Rules := []*Rule{}
+	if len(LeaderConstraints) > 0 {
+		Rules = append(Rules, NewRule(Leader, 1, LeaderConstraints))
+	}
+
+	if voterCount > 0 {
+		VoterRules, err := NewRules(Voter, voterCount, voterConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid VoterConstraints", err)
+		}
+		for _, rule := range VoterRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: VoterConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, VoterRules...)
+	}
+
+	if followerCount > 0 {
+		FollowerRules, err := NewRules(Follower, followerCount, followerConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid FollowerConstraints", err)
+		}
+		for _, rule := range FollowerRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: FollowerConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, FollowerRules...)
+	}
+
+	if learnerCount > 0 {
+		LearnerRules, err := NewRules(Learner, learnerCount, learnerConstraints)
+		if err != nil {
+			return nil, fmt.Errorf("%w: invalid LearnerConstraints", err)
+		}
+		for _, rule := range LearnerRules {
+			for _, cnst := range CommonConstraints {
+				if err := rule.Constraints.Add(cnst); err != nil {
+					return nil, fmt.Errorf("%w: LearnerConstraints conflicts with Constraints", err)
+				}
+			}
+		}
+		Rules = append(Rules, LearnerRules...)
+	}
+
+	return &Bundle{Rules: Rules}, nil
```

Comment:
Do we need to set `override` a true value to override default group?

Reply:
It will be reset when invoking `Reset`, that API will be called almost everywhere. It is more suitable as the only entry/validation, since we have multiple new methods.